asn1/CommonLedgerProtocol.asn

```diff
@@ -0,0 +1,114 @@
+CommonLedgerProtocol
+DEFINITIONS
+AUTOMATIC TAGS ::=
+BEGIN
+
+IMPORTS
+    UInt8,
+    UInt32,
+    UInt64,
+    UInt128,
+    UInt256
+    FROM GenericTypes
+
+    Timestamp
+    FROM InterledgerTypes
+
+    InterledgerProtocolError
+    FROM InterledgerProtocol
+;
+
+-- Zero or more protocolData entries, each with a unique
+-- protocolName, describe protocol-specific information about
+-- unpaid, paid, and conditionally paid requests and results
+-- sent over the CommonLedgerProtocol.
+-- The protocolName of each entry describes the protocol for
+-- interpreting the data. Peers need to establish beforehand
+-- which protocols each understand, and which numbers they
+-- use, which can be agreed ad-hoc,except that protocolName 'ilp'
+-- is reserved for the Interledger protocol.
+-- When multiple protocolData items are sent in a request,
+-- only the first one will be decisive in triggering the
+-- response, the rest of the protocolData items are considered
+-- side protocols that piggyback on the first one.
+-- The contentType byte can be used to allow debug tools to
+-- display the contents of the protocolData without understanding
+-- it.
+
+ContentType ::= INTEGER {
+  applicationOctetString  (0),
+  textPlainUtf8           (1),
+  applicationJson         (2)
+} (0..255)
+
+ProtocolData ::= SEQUENCE OF SEQUENCE {
+  protocolName IA5String,
+  contentType ContentType,
+  data OCTET STRING
+}
+
+-- Ack, Response, and Error are the response types.
+-- When using these in a CLP packet, the requestId should match
+-- the requestId of the request they respond to.
+
+Ack ::= ProtocolData
+Response ::= ProtocolData
+Error ::= SEQUENCE {
+  rejectionReason InterledgerProtocolError,
+  --
+  protocolData ProtocolData
+}
+
+-- Prepare, Fulfill, Reject, and Message are the request types.
+-- Each request should have a unique requestId.
+
+Prepare ::= SEQUENCE {
+  transferId UInt128,
+  amount UInt64,
+  executionCondition UInt256,
+  expiresAt Timestamp,
+  --
+  protocolData ProtocolData
+}
+
+Fulfill ::= SEQUENCE {
+  transferId UInt128,
+  fulfillment UInt256,
+  --
+  protocolData ProtocolData
+}
+
+Reject ::= SEQUENCE {
+  transferId UInt128,
+  rejectionReason InterledgerProtocolError,
+  --
+  protocolData ProtocolData
+}
+
+Message ::= ProtocolData
+
+CALL ::= CLASS {
+    &typeId UInt8 UNIQUE,
+    &Type
+} WITH SYNTAX {&typeId &Type}
+
+CallSet CALL ::= {
+    {1 Ack} |
```

Comment:
Ack and Response both indicate a success here, and both have the same definition. Can't we use empty Responses in place of a special Ack type?

Reply:
Yes, I think so. I am almost done with a first implementation and I don't use `Ack` at all.

Reply:
Revising my earlier examples:

(ProtocolData in parenthesis.)

#### Quote (successful)

``` lua
>>> Message (ilp => 0x02...) -- TYPE_ILQP_LIQUIDITY_REQUEST
<<< Response (ilp => 0x03...) -- TYPE_ILQP_LIQUIDITY_RESPONSE 
```

#### Quote (locally successful, but connector responded with an ILP error)

``` lua
>>> Message (ilp => 0x02...) -- TYPE_ILQP_LIQUIDITY_REQUEST
<<< Response (ilp => 0x08...) -- TYPE_ILP_ERROR
```

#### Quote (local error)

``` lua
>>> Message (ilp => 0x02...) -- TYPE_ILQP_LIQUIDITY_REQUEST
<<< Error (...) -- Some CLP error format we define
```


#### Successful payment

``` lua
>>> Prepare (ilp => 0x01...) -- TYPE_ILP_PAYMENT
<<< Response () -- Empty response [for now?]

<<< Fulfill ()
>>> Response ()
```

#### Failed payment

``` lua
>>> Prepare (ilp => 0x01...) -- TYPE_ILP_PAYMENT
<<< Response ()

<<< Reject ()
>>> Response ()
```

Some questions:

* Should the `Reject` message contain a rejectionReason field (as the ASN says currently) or just an ILP Error in the ProtocolData?
* Should we define a `Fulfill` format on the ILP layer which may, for example, communicate some data back to the payer? (see #279)

Reply:
moving this to #282